### DISCUSSION TITLE
Managing Xcode at scale

### DISCUSSION THREAD

**Discussion by ecamacho** [MobileNativeFoundation/discussions#35] [General]
Hi all!

So I’m wondering how you handle Xcode at scale. Specifically, **how are you bumping Xcode in CI and local environments, and how often do you do it?**

We, at Spotify, bump to major versions ASAP. That usually means around October, when those versions are released. For minor versions, we bump to them if a team requests it (maybe they need it for some feature they are working on) or if there are some important bugs fixed.

## Our process

1. We manually download and unxip the Xcode app. We gzip it and upload it to our infra (Google Cloud Storage).
1. We run our CI jobs with the new version to verify everything is OK. 
1. We fix some issues ourselves in the Platform teams, but sometimes we file tickets to other teams to help us for specific issues such as broken Smoke tests. 
1. We also monitor build time regressions using XCMetrics.
1. This process can be as quick as one week or as long as a couple of months, specially for major versions or if we also need to bump MacOS.
1. Once everything is OK, we bump the Xcode version by editing an `.xcode-version` file at the root of our projects.

## Our tooling

We have tools that select the right Xcode version based on the`.xcode-version` file. That setting can be overridden using an environment variable.

- In **CI**, we download the Xcode version needed, if is not already present in the build agent local cache, and setup the `DEVELOPER_DIR` env variable to point to it.
- In **local machines**, we present a warning if the user is using a different minor version or an error if she’s using a different major version and prompt her to download it from our internal infra.

## Things that we want to improve:

1. Remove the manual steps to download /upload and run CI jobs with the new Xcode version
1. Provide an automated way for local devs to install Xcode
1. We’re also thinking about blocking developers to install Xcode from the AppStore, since those versions are automatically upgraded leading to build errors, specially for new major versions.

**Comment by alephao** [comment]
Hi @ecamacho, can I ask what is your CI set up? Do you use a cloud provider or in-house? If it's in-house, do you use virtualization?

**Reply by BalestraPatrick** [reply]
@alephao We touched on those points in this other thread: https://github.com/MobileNativeFoundation/discussions/discussions/23#discussioncomment-424955

**Comment by ccarpita** [comment]
At [Butterfly Network](https://www.butterflynetwork.com/index.html) we partially automate the process of updating and installing Xcode versions, and try to keep up-to-date with all minor release versions to ensure developers can upgrade their iPhones in a reasonable timeframe. Our CI agents have multiple versions of Xcodes pre-provisioned, which stands in contrast to the JIT provisioning that I know Spotify did with its vulcan tool. This design choice is mostly an artifact of disk limitations on our build agents, and a preference to keep build timings consistent.

Developers can run a rake command in the form `rake provision:xcode`, which will download a `xip` from our AWS bucket, unpack it, and install it to the Applications directory in the form `/Applications/Xcode-v(version).app`, and then verify with `spctl`. Most of the provisioning steps are specified in Ansible, and we would likely have a hard time open-sourcing this part of our automation.

The software-infra team owns a single bespoke script that can used to list releases using the [XcodeReleases](https://xcodereleases.com) API and initiate support for a release, which entails prompting to download the `xip` along with CLI-tools, uploading the files to our AWS bucket, and adding build-version, sha256 hashes, and other details to an Ansible role variable file, thus allowing deployment to CI or local installation by developers.

The one thing holding us back from automating Xcode deployments, more than anything, is Apple's aggressive 2FA policies for Xcode downloads. We've looked into tools like [xcodes](https://github.com/RobotsAndPencils/xcodes) to manage this, but have had persistent issues actually getting them to work consistently. We'd definitely like to hear if folks have worked around this issue!

**Comment by keith** [comment]
At Lyft we generally bump major versions ASAP as well. We also often do the same with the minor versions, although there were some scares this year with some bugs that held us back a little bit.

Our process:

1. We manually download the xip, and upload it to s3
2. Test it locally as needed
3. Update a few CI machines to the new version
4. Push a branch with the required version bump and other changes to run against the new CI machines
5. Notify developers a few days ahead of time so they can download the new version and be ready for when we swap over
6. Merge it after our release cut to give us the maximum amount of time to find unexpected regressions

We enforce this with a local script that checks the output of `xcodebuild -version` against an allowlist and error when developers are generating a new project (which is a required step in our workflow) that they need to download the right version. We also [lock the versions in bazel](https://www.smileykeith.com/2021/03/08/locking-xcode-in-bazel/) since there are some paths you can get around that script with and we don't want people to accidentally get cache misses.

I think our desired improvements are similar to those you posted, we don't stop people from using the MAS for it today, but that causes a ton of issues when people get auto-updated and then have to go download a previous version somewhere. We also let devs download from a s3 bucket rather than having to login to the portal, but that's still not super automatic. Sounds like what @ccarpita does is a nice improvement on some of this

**Comment by aaroncrespo** [comment]
On Outlook iOS we lean on ADO for a lot of this but have also built provisioning tooling around it to streamline it further. Provisioning here means setting up and verifying: Xcode, homebrew/Brewfile dependencies, certificates, git configuration, minimal python tooling etc

We Bump Major Versions ASAP. For minor and patch versions we also try to get them in ASAP despite some surprises this year with Big Sur and Build System Changes.

___
CI and Hardware:

We use ADO build machines which typically get Xcode Releases and Betas on weekly drops. Which is fine at worse an awkwardly timed release will cause a 2 week gap. We don't have to worry to much about hardware/os/Xcode versions that way.

For machine setup we don't need to do much here. ADO pipelines let us clone a parallel set of pipeline and Xcode/OS versions are parameterized. This lets us create a Beta toolchain + pipeline in parallel to our Prod pipeline for as long as a transition is needed. Its all copying build definitions and a few changes to a config and yaml file. 

ADO pipelines are similar to Github Actions.
___
Provisioning and Xcode Management:

Our provisioning tooling is as automated as we can make it. Developers can run it locally to verify their setup, or use it to bootstrap it - CI always runs it.

The tooling checks/gets our dependencies, but more importantly it verifies the correct Xcode is selected (uses the same config that parametrizes the CI pipeline vm selection). It's not expected that developers need to be constantly verifying their configs, but it is expected that if the developer sees a build issue that they then verify their setup. This allows developers a couple days of lag time to download Xcode versions or finish old branches when time/bandwidth is best for them. We try to have it so that the tooling can "fix-it" most build environment issues for local dev machines or ci machines.

The verification pass is very quick and was inspired by `brew doctor` however,  since we can be opinionated it can fix issues it finds, not just highlight them

Theres a "contract" where if `doctor` passes verification, then your build should compile and run baring any local changes. This has saved a lot of time on both build and feature teams.

**Comment by alanzeino** [comment]
I built this process during my time at Uber, generally:
1. Version control your current Xcode version, so you always know what version of the toolchain is/was needed for any given git ref in your repo, makes snapshot and unit testing easier
2. Set up your infrastructure to be able to dynamically switch the Xcode version inside the CI job at the start on any given CI node (i.e., mac) and shake out any bugs related to this switching like blowing away caches or killing processes before or afterwards so that this is stable
3. Make it easy for engineers to download, unpack and install Xcode with a command line tool they can run
4. Don't distribute Xcode via Munki or Chef, just don't
5. Don't allow App Store versions of Xcode to be used for the same reason as point `4` which is that you never want the Xcode version to randomly change behind a developer's back
6. And my personal favourite (but optional), if you have multiple apps inside your company you should probably use the same Xcode version for all of them, otherwise if you let some apps use different versions you're going to have some headaches with testing shared libraries

**Comment by JoeSSS** [comment]
At XING we use Anka for provisioning of the CI, so we manually update the Xcode version and create a new template which then gets distributed to all the executors. 
We do not lock Xcode on developer machines although it might be a good idea.
